### ISSUE TITLE
Find a solution for storing and using secrets in production

### ISSUE THREAD

**Issue by humphd** [Seneca-CDOT/telescope#2979]
In #2978 we're applying a bandaid fix for putting secrets into our production containers via the autodeployment server, by injecting the `.env` file with new values.

This is a good temporary solution, but there are better ways to do this with Docker Swarm secrets or something like Vault, etc.  Let's look for a long term solution to this so that we can keep encrypted secrets somehow.

**Comment by humphd** [comment]
It looks to me like we could try to [use `docker secret`](https://docs.docker.com/engine/swarm/secrets/#read-more-about-docker-secret-commands) for our staging and production secrets.

It works like this:

1. You have some secret, let's say our `GITHUB_TOKEN`, and let's pretend the value is `abcd`.  You start by creating a new secret with `docker` on the staging and production boxes like this: `printf "abcd" | docker secret create github_token -`
2. You can see the secrets (by name, not their value) using `docker secret ls` and use `docker secret rm <secret_name>` to remove.
3. In our containers, these secrets would get mounted to a read-only, temporary filesystem at `/run/secrets/<secret_name>`, so our GitHub Token would be available at `/run/secrets/github_token`.
4. We set up the secret in a container using [docker-compose like this](https://docs.docker.com/engine/swarm/secrets/#use-secrets-in-compose).
5. We need to implement some method for turning the secret file(s) into environment variables.  People seem to do it various ways, for example: [overriding the `ENTRYPOINT` to call a shell script that does it](https://stackoverflow.com/questions/48094850/docker-stack-setting-environment-variable-from-secrets).  Here's a sample repo that shows the full method: https://github.com/rbdiang/docker-secrets-example

I think this solves a bunch of our problems, and doesn't require us to use a new secrets container technology.

To summarize: we define our secrets on staging and production and modify our `docker-compose` to read secrets files at runtime into environment variables that we can use inside the containers.

@manekenpix you cool with this?

Getting this done soon is pretty important, since a number of our deployments depend on secrets now (Supabase as one example).

**Comment by humphd** [comment]
Filed https://github.com/supabase/supabase/issues/6014 upstream to see if Supabase would consider using secrets files vs. environment variables.

Even if they do, it wouldn't happen in time for what we need, so let's move forward with the environment variable hack, and hopefully we can migrate away from it later.

**Comment by joelazwar** [comment]
gotcha. I've got a bash script down to convert the secrets to env variables. But I'm a little stuck on our docker-compose.

So for each service that uses JWT_SECRET, we'll need to override their entrypoint to run the said script?

**Comment by humphd** [comment]
It turns out that we can only embed secrets in docker running in swarm mode.  Brutal.

We're going to do something else, and use a separate `secrets.env` file to hold them which lives outside of git.

@joelazwar is going to work up a solution based on a conversation in Slack.
